lib/Widgets/SeekBar.vala

```diff
@@ -16,34 +16,232 @@
+
+    /*
+     * Emitted when a scroll action is performed on the scale.
+     */
+    public signal void scroll_action (Gtk.ScrollType scroll, double new_value);
+
+    /*
+     * Emitted when the region has been allocated for the scale.
+     */
+    //public signal void scale_size_allocate (Gtk.Allocation alloc_rect);
+
+    /*
+     * Creates a new SeekBar with a fixed playback duration.
+     * */
+    public SeekBar (double playback_duration) {
+        Object (playback_duration: playback_duration);
     }
 
     construct {
-        var progression_label = new Gtk.Label ("0:00");
+        get_style_context ().add_class ("seek-bar");
 
-        var scale = new Gtk.Scale.with_range (Gtk.Orientation.HORIZONTAL, 0, 1, 0.1);
-        scale.draw_value = false;
+        /* GUI */
+        orientation = Gtk.Orientation.HORIZONTAL;
+        column_spacing = 6;
+        halign = Gtk.Align.CENTER;
+        progression_label = new Gtk.Label ("");
+        progression_label.get_style_context ().add_class ("progression-label");
+        time_label = new Gtk.Label ("");
+        time_label.get_style_context ().add_class ("time-label");
+        progression_label.margin_right = time_label.margin_left = 3;
+
+        scale = new Gtk.Scale.with_range (Gtk.Orientation.HORIZONTAL, 0, 1, 0.1);
+        scale.get_style_context ().add_class ("scale");
         scale.hexpand = true;
+        scale.draw_value = false;
+        scale.can_focus = false;
+        scale.events |= Gdk.EventMask.POINTER_MOTION_MASK;
+        scale.events |= Gdk.EventMask.LEAVE_NOTIFY_MASK;
+        scale.events |= Gdk.EventMask.ENTER_NOTIFY_MASK;
+
+        is_grabbing = false;
+        is_hovering = false;
+
+        set_duration (playback_duration);
+        set_progress (0);
+
+        /* signals */
+        scale.button_press_event.connect ((event) => {
+            is_grabbing = true;
+            scale_button_press (event);
+            return false;
+        });
+
+        scale.button_release_event.connect ((event) => {
+            is_grabbing = false;
+            set_progress (scale.get_value ());
+            scale_button_release (event);
+            return false;
+        });
 
-        var time_label = new Gtk.Label ("0:00");
+        scale.enter_notify_event.connect ((event) => {
+            is_hovering = true;
+            scale_hover (event);
+            return false;
+        });
+
+        scale.leave_notify_event.connect ((event) => {
+            is_hovering = false;
+            scale_leave (event);
+            return false;
+        });
+
+        scale.motion_notify_event.connect ((event) => {
+            set_progress (scale.get_value ());
+            scale_motion (event);
+            return false;
+        });
+
+        scale.change_value.connect ((scroll, new_value) => {
+            scroll_action (scroll, new_value);
+            return false;
+        });
+
+        scale.size_allocate.connect ((alloc_rect) => {
```

Comment:
This connection seems to do nothing in it's lambda. You could just remove it altogheter.